File: Day1-9/4.go
package main

import (
	"fmt"
	"strconv"
)

func isSorted(s string) bool {
	for i := 1; i < len(s); i++ {
		if s[i] < s[i-1] {
			return false
		}
	}
	return true
}

func passwordGuess(a, b int) (int, int) {
	p1, p2 := 0, 0
	for x := a; x <= b; x++ {
		xStr := strconv.Itoa(x)
		if !isSorted(xStr) {
			continue
		}
		digits := make(map[byte]int)
		for i := 0; i < len(xStr); i++ {
			digits[xStr[i]]++
		}
		for _, v := range digits {
			if v >= 2 {
				p1++
				break
			}
		}
		for _, v := range digits {
			if v == 2 {
				p2++
				break
			}
		}
	}
	return p1, p2
}

func main() {
	p1, p2 := passwordGuess(236491, 713787)

	fmt.Printf("Part 1: %d\n", p1)
	fmt.Printf("Part 2: %d\n", p2)
}

File: Day1-9/4.py
from collections import Counter

A = 236491
B = 713787


def password_count():
    valid_p1 = 0
    valid_p2 = 0
    for x in range(A, B):
        x_str = str(x)
        if ''.join(sorted(x_str)) != x_str:
            continue

        counter = Counter(x_str)
        if any(v for v in counter.values() if v >= 2):
            valid_p1 += 1
        if any(v for v in counter.values() if v == 2):
            valid_p2 += 1

    return valid_p1, valid_p2


if __name__ == "__main__":
    p1, p2 = password_count()
    print("Part 1: {}".format(p1))
    print("Part 2: {}".format(p2))
